uaclient/util.py

```diff
@@ -801,22 +801,22 @@ def parse_rfc3339_date(dt_str: str) -> datetime.datetime:
 
     :return: datetime.datetime object of time represented by dt_str
     """
+    # remove sub-seconds
+    dt_str_without_subseconds = re.sub(
+        r"(\d{2}:\d{2}:\d{2})\.\d+", r"\g<1>", dt_str
```

Comment:
Can't we just combine  `dt_str_without_subseconds` with `dt_str_with_z` by using:

```python
    dt_str_without_subseconds = re.sub(
        r"(\d{2}:\d{2}:\d{2})\.\d+", r"\g<1>Z", dt_str)
```

Also, I may be missing something, but are we changing the format of the output string now in the end ? If we are, can't this affect already stored datetime strings in the system ?

Reply:
> Can't we just combine `dt_str_without_subseconds` with `dt_str_with_z` by using:
> 
> ```python
>     dt_str_without_subseconds = re.sub(
>         r"(\d{2}:\d{2}:\d{2})\.\d+", r"\g<1>Z", dt_str)
> ```
> 

We totally could combine them, but I kept them separate solely for readability. Because regexes are hard enough to come back to 6 months later and read already, I figure that keeping each step with a single purpose would help.


> Also, I may be missing something, but are we changing the format of the output string now in the end ? If we are, can't this affect already stored datetime strings in the system ?

This is just a parsing function, so this won't affect how we're formatting datestrings during serialization - that is done at `util.py:105`. The adjustment to this function just adds support for parsing another variant of a datestring.

Before this change, it was failing to parse a datestring with subseconds, but without a timezone, which is why there had to be some ad hoc date parsing in `timer.py`. But by adding support for this format, we can just rely on DateTimeAwareJsonDecoder to parse dates in this format. I added this format to the unit test cases for this function at `test_util.py:934`

Reply:
Okay, good points and I do agree that reading regex is a though call. One thing we can do make that easier in the feature is having some code docs with a string before and after applying the regex.

Reply:
Good idea! I added some comments with examples